### PR TITLE
20 sharedsession

### PR DESCRIPTION
**Pull request by timjon**

### CODE REVIEW

File: Application/src/controller/network/Decode.java [hunk 1]
```diff
@@ -66,7 +67,11 @@ void execute() {
             // Writes the rawStringToDecode data
             write(id, "INFO: " + retrieveMessage(rawStringToDecode));
 
-            // If it's a message or a list of classes (new diagram).
+            //Info messages
+        } else if (rawStringToDecode.contains("INFO#")) {
+
+            disp("Info", rawStringToDecode.substring(6), "");
```
Comment (pontuslaestadius) [inline Application/src/controller/network/Decode.java]: Why does it substring at 6? Please comment I'm retarded.

File: Application/src/controller/network/Decode.java [hunk 2]
```diff
@@ -95,6 +100,7 @@ void execute() {
                 diagramClasses(message, id);
             }
         }
+        Menu.getInstance().identifyState();
```
Comment (pontuslaestadius) [inline Application/src/controller/network/Decode.java]: Add a comment for it!
something like: This is here so proceeding all decoding the button states are updated.

Comment (pontuslaestadius) [reply Application/src/controller/network/Decode.java]: Couldn't this be placed elsewher?

File: Application/src/model/Menu.java [hunk 3]
```diff
@@ -3,8 +3,19 @@
...
+import javafx.event.EventHandler;
+import javafx.geometry.Insets;
+import javafx.geometry.Pos;
+import javafx.scene.control.Button;
+import javafx.scene.control.Label;
+import javafx.scene.control.PasswordField;
+import javafx.scene.control.TextField;
+import javafx.scene.layout.GridPane;
+import javafx.scene.paint.Color;
+import javafx.scene.text.Text;
```
Comment (pontuslaestadius) [inline Application/src/model/Menu.java]: It would look cleaner if you important scene.*;

File: Application/src/model/Menu.java [hunk 4]
```diff
@@ -20,6 +31,9 @@
  */
 public class Menu {
 
+    private static boolean host = false;
```
Comment (pontuslaestadius) [inline Application/src/model/Menu.java]: Please comment here

File: Application/src/model/Menu.java [hunk 5]
```diff
@@ -20,6 +31,9 @@
...
+    private static boolean joined_lobby = false;
```
Comment (pontuslaestadius) [inline Application/src/model/Menu.java]: This should need a comment!

File: Application/src/model/Menu.java [hunk 6]
```diff
@@ -28,6 +42,11 @@
     private static Button button_next;
     private static Button button_previous;
     private static Button button_auto;
+    private static Button button_join_lobby;
+    private static Button button_create_lobby;
+    private static Button button_remove_lobby;
+    private static Button button_import_lobby;
+    private static Button button_leave_lobby;
```
Comment (pontuslaestadius) [inline Application/src/model/Menu.java]: These aren't in the order we use elsewhere? Like how we add children,

File: Application/src/model/Menu.java [hunk 7]
```diff
@@ -37,6 +56,11 @@
     private final static String text_auto_pause = " || ";
     private final static String text_next = "->";
     private final static String text_previous = "<-";
+    private final static String text_import_lobby = "Import to lobby";
+    private final static String text_join_lobby = "Join a lobby";
```
Comment (pontuslaestadius) [inline Application/src/model/Menu.java]: Why is this the only one that uses 'a' join A lobby? Please use the same standard for all of them.

File: Application/src/model/Menu.java [hunk 8]
```diff
@@ -209,25 +244,151 @@ private void setEvents(Stage stage) {
                         break;
                 }
 
-
             }
 
             identifyState();
         });
 
+        button_import_lobby.setOnAction((ActionEvent event) -> {
```
Comment (pontuslaestadius) [inline Application/src/model/Menu.java]: This button has a lot of copy paste from button Import

Comment (pontuslaestadius) [reply Application/src/model/Menu.java]: So, fix all of that by making functions. K ty.

File: Application/src/model/Menu.java [hunk 9]
```diff
@@ -209,25 +244,151 @@ private void setEvents(Stage stage) {
...
+
+                        // Catches if there are no diagrams.
+                        try {
+                            Net.push("{share, " + parse.getDiagram() + "}");
+                            System.out.println("{share, " + parse.getDiagram() + "}");
+                        } catch (NullPointerException e) {
+                            continue;
+                        }
+
+                        // Catches if there is no parallel diagram.
```
Comment (pontuslaestadius) [inline Application/src/model/Menu.java]: This comment does not work?

File: Application/src/model/Menu.java [hunk 10]
```diff
@@ -209,25 +244,151 @@ private void setEvents(Stage stage) {
...
-            Net.push("{" + DiagramView.getDiagramViewInView().getTab().getId() + ", previous_message}");
+            String tabID = DiagramView.getDiagramViewInView().getTab().getId();
+            try {
+                //Checks if the current tab has a diagram from a lobby or not
+                if (tabID.charAt(1) == 'l') {
+                    Net.push("{share, {" + tabID + ", previous_message}}");
+                } else {
+                    Net.push("{" + tabID + ", previous_message}");
+                }
+            }catch(StringIndexOutOfBoundsException e){
```
Comment (pontuslaestadius) [inline Application/src/model/Menu.java]: Comment why you catch

File: Application/src/model/Menu.java [hunk 11]
```diff
@@ -209,25 +244,151 @@ private void setEvents(Stage stage) {
...
-            Net.push("{" + DiagramView.getDiagramViewInView().getTab().getId() + ", next_message}");
+            String tabID = DiagramView.getDiagramViewInView().getTab().getId();
+            try {
+                //Checks if the current tab has a diagram from a lobby or not
+                if (tabID.charAt(1) == 'l') {
+                    Net.push("{share, {" + tabID + ", next_message}}");
+                } else {
+                    Net.push("{" + tabID + ", next_message}");
+                }
+            }catch(StringIndexOutOfBoundsException e){
```
Comment (pontuslaestadius) [inline Application/src/model/Menu.java]: Add a comment here

File: Application/src/model/Menu.java [hunk 12]
```diff
@@ -209,25 +244,151 @@ private void setEvents(Stage stage) {
...
+                Net.push("{" + tabID + ", next_message}");
+            }
         });
 
         button_auto.setOnAction((ActionEvent event)    ->{
             play = button_auto.getText().endsWith(text_auto_play);
             identifyState();
         });
 
+        button_create_lobby.setOnAction((ActionEvent event)    ->{
```
Comment (pontuslaestadius) [inline Application/src/model/Menu.java:234]: Why do you call a 'passwordBox'?

File: Application/src/model/Menu.java [hunk 13]
```diff
@@ -209,25 +244,151 @@ private void setEvents(Stage stage) {
...
+            passwordBox();
+            identifyState();
+        });
+        button_remove_lobby.setOnAction((ActionEvent event) ->{
+            Net.push("{" + "share, " + "remove_lobby}");
+
+            host = false;
```
Comment (pontuslaestadius) [inline Application/src/model/Menu.java]: Now you are assuming that you can do this?

File: Application/src/model/Menu.java [hunk 14]
```diff
@@ -209,25 +244,151 @@ private void setEvents(Stage stage) {
...
+            identifyState();
+        });
+
+        button_join_lobby.setOnAction((ActionEvent event)    ->{
+            //Creates the popup window
+            Stage joinLobbyStage = new Stage();
+            joinLobbyStage.setTitle("Join a lobby");
+
+
```
Comment (pontuslaestadius) [inline Application/src/model/Menu.java]: Please remove an empty line thank you

File: Application/src/model/Menu.java [hunk 15]
```diff
@@ -209,25 +244,151 @@ private void setEvents(Stage stage) {
...
+                    VBox vbox = new VBox();
+
+                    TextField lobbyID = new TextField("");
+                    lobbyID.setPromptText("Lobby ID");
+                    lobbyID.setFocusTraversable(false);
```
Comment (pontuslaestadius) [inline Application/src/model/Menu.java]: What does this do?

File: Application/src/model/Menu.java [hunk 16]
```diff
@@ -209,25 +244,151 @@ private void setEvents(Stage stage) {
...
+
+                    PasswordField password = new PasswordField();
```
Comment (pontuslaestadius) [inline Application/src/model/Menu.java]: Why do you use a password  field? plz commentzxzxc

File: Application/src/model/Menu.java [hunk 17]
```diff
@@ -209,25 +244,151 @@ private void setEvents(Stage stage) {
...
+            vbox.getChildren().add(join_button);
+            vbox.getChildren().add((label));
+
+            //Sets the size of the window
+            Scene stageScene = new Scene(vbox, 300, 100);
+            joinLobbyStage.setScene(stageScene);
+            joinLobbyStage.show();
+
+
+                    join_button.setOnAction((ActionEvent e) -> {
```
Comment (pontuslaestadius) [inline Application/src/model/Menu.java:280]: Nested Event handler??????????????????????????????????????????????????????? 
ENd me plz.

File: Application/src/model/Menu.java [hunk 18]
```diff
@@ -209,25 +244,151 @@ private void setEvents(Stage stage) {
...
+                        String id = lobbyID.getText();
+                        String pwd = password.getText();
+
+                        if (id.equals("") || pwd.equals("")) {
```
Comment (pontuslaestadius) [inline Application/src/model/Menu.java]: why? commentt tyty.

File: Application/src/model/Menu.java [hunk 19]
```diff
@@ -209,25 +244,151 @@ private void setEvents(Stage stage) {
...
+                        } else {
+                            Net.push("{share, {join_lobby, {" + lobbyID.getText() + ", " + password.getText() + "}}}");
+                            // Joined lobby
+                            joined_lobby = true;
+                            joinLobbyStage.close();
+                        }
+                    });
+
+                });
+        button_leave_lobby.setOnAction((ActionEvent event) ->{
```
Comment (pontuslaestadius) [inline Application/src/model/Menu.java]: give this some more room it looks hidden!

File: Application/src/model/Menu.java [hunk 20]
```diff
@@ -209,25 +244,151 @@ private void setEvents(Stage stage) {
...
+            leaveLobbyBox();
```
Comment (pontuslaestadius) [inline Application/src/model/Menu.java]: what this do?

File: Application/src/model/Menu.java [hunk 21]
```diff
@@ -287,24 +447,53 @@ public void identifyState() {
             pause();
         }
 
+        Platform.runLater(() -> {
```
Comment (pontuslaestadius) [inline Application/src/model/Menu.java:361]: This is some really amazing code! How did you come to this conclusion that it would work?

File: Application/src/model/Menu.java [hunk 22]
```diff
@@ -287,24 +447,53 @@ public void identifyState() {
...
         } catch (IllegalStateException ex) {
-
-            // If there is no view, we disable the media buttons.
-            Menu.getInstance().setMenuState(false, button_import);
             return;
         }
 
-        // Updates optional view states.
-        ArrayList<String> viewing = diagramView.getViewing();
+        String id = diagramView.getTab().getId();
```
Comment (pontuslaestadius) [inline Application/src/model/Menu.java]: why is this added here?

File: Application/src/model/Menu.java [hunk 23]
```diff
@@ -338,4 +526,88 @@ public void identifyState() {
         });
     }
 
+    /**
+     * Method for adding a password to the creation of a lobby.
```
Comment (pontuslaestadius) [inline Application/src/model/Menu.java]: I know it's a method. I'm not retarded

File: Application/src/model/Menu.java [hunk 24]
```diff
@@ -338,4 +526,88 @@ public void identifyState() {
...
+     */
+    public void passwordBox(){
+        String ok_password_text = "OK";
```
Comment (pontuslaestadius) [inline Application/src/model/Menu.java]: Maybe use text first? 
e.g. text_password_ok

File: Application/src/model/Menu.java [hunk 25]
```diff
@@ -338,4 +526,88 @@ public void identifyState() {
...
+        Label pw = new Label("Password:"); // Label showing showing next to textField.
+        grid.add(pw, 0, 0);
+
+        PasswordField pwBox = new PasswordField(); // Textfield with password protection.
+        grid.add(pwBox, 1, 0);
+        pwBox.setPromptText("Password");
+        Button button_ok_password = new Button(ok_password_text); // Button to create a lobby with input password.
+        grid.add(button_ok_password, 1,1);
+
+        final Text pressReturn = new Text(); //
```
Comment (pontuslaestadius) [inline Application/src/model/Menu.java]: // you too

File: Application/src/model/Menu.java [hunk 26]
```diff
@@ -338,4 +526,88 @@ public void identifyState() {
...
+                    pressReturn.setFill(Color.GREEN);
+                    pressReturn.setText("Lobby created");
+                    host = true;
+                    primaryStage.close();
+                }
+            }
+        });
+    }
+
+    public void leaveLobbyBox(){
```
Comment (pontuslaestadius) [inline Application/src/model/Menu.java]: No javadocs

Comment (pontuslaestadius) [inline Application/src/model/Menu.java]: No javadocs

**Review by pontuslaestadius — CHANGES_REQUESTED** [review]
I did a quick overview and found a few faults which needs to be corrected //

File: Application/Application.iml [hunk 27]
```diff
@@ -8,14 +8,5 @@
     <orderEntry type="inheritedJdk" />
     <orderEntry type="sourceFolder" forTests="false" />
     <orderEntry type="library" name="gson-2.8.2" level="project" />
-    <orderEntry type="module-library" exported="">
-      <library>
-        <CLASSES>
-          <root url="jar://$MODULE_DIR$/lib/gson-2.8.2.jar!/" />
```
Comment (ghost) [inline Application/Application.iml:14]: pls undo these deletions

File: Application/src/model/Menu.java (hunk 21 above)
Comment (Sebastian-Fransson) [reply Application/src/model/Menu.java:361]: I asked a magician that lives on the fourth floor in Patricia. He directed me to a hidden expert that showed me the way to enlightenment. Then I asked Pontus about the buttons and he told me to add it there.

File: Application/src/model/Menu.java (hunk 12 above)
Comment (Sebastian-Fransson) [reply Application/src/model/Menu.java:234]: Would it be better with passwordWindow oe passwordInput? I call it that because it is a textbox for passwords.

File: Application/src/model/Menu.java (hunk 13 above)
Comment (Sebastian-Fransson) [reply Application/src/model/Menu.java]: ?

File: Application/src/model/Menu.java [hunk 28]
```diff
@@ -338,4 +433,159 @@ public void identifyState() {
...
+                    leaveStage.close();
+                }
+        });
+    }
+
+    /**
+     * Sends a message of the given message type to the server
+     * @param MessageType The message type, either previous_message or next_message
+     */
+    public static void changeMessage(String MessageType){
```
Comment (pontuslaestadius) [inline Application/src/model/Menu.java]: ChangeMessage doesn't sound like a class which sends messages to the server. Maybe rename it and move in to the network class too lower menu coupling?

File: Application/src/model/Menu.java [hunk 29]
```diff
@@ -338,4 +433,159 @@ public void identifyState() {
...
+        } else {
+            Net.push("{" + tabID + ", " + MessageType + "}");
+        }
+    }
+
+    /**
+     * Handels importing a diagram
+     * @param share True if the diagram is supposed to be shared othervise false
+     */
+    public void importDiagram(Stage stage, Boolean share) {
```
Comment (pontuslaestadius) [inline Application/src/model/Menu.java]: This could probably be in the import.java file and and doesn't need to be in menu.java?

File: Server/src/diagramcoordinator.erl [hunk 30]
```diff
@@ -55,15 +55,15 @@ loop(Usercoordinator, Did, Pids, [], Message_number, PrevList) ->
...
+    {next_message, Coordinator} -> 
+	  Coordinator ! ok,
       {From, To, Message} = L,
-	  send_message(find_pid(Pids, From), From, To, Message, find_pid(Pids, To), Message_number, Usercoordinator, Did), 
+	  send_message(find_pid(Pids, From), From, To, Message, find_pid(Pids, To), Message_number, Coordinator, Did), 
       receive
 	    {message_done, From, To, Message, Message_number} ->
-		  Usercoordinator ! {message_sent, Did, From, To, Message, Message_number},
+		  Coordinator ! {message_sent, Did, From, To, Message, Message_number},
 		  %Sends info to the usercoordinator that a message has been received by a node. To be printed in the executionlog
```
Comment (pontuslaestadius) [inline Server/src/diagramcoordinator.erl]: The comments in this file has not been updated. they all say usercoordinator

File: Server/src/loadbalancer.erl [hunk 31]
```diff
@@ -2,15 +2,18 @@
...
   loop(ListenSocket).
   
 %Accepts connections and spawns a proccess for it
 loop(ListenSocket) ->
   {ok, Socket} = gen_tcp:accept(ListenSocket),
-  spawn(fun() -> loop(ListenSocket) end),
-  usercoordinator:init(Socket).
\ No newline at end of file
+  Pid = spawn(fun() -> usercoordinator:init(Socket) end),
+  gen_tcp:controlling_process(Socket, Pid),
```
Comment (pontuslaestadius) [inline Server/src/loadbalancer.erl]: I don't know what this line means, please add a comment

File: Server/src/lobby.erl [hunk 32]
```diff
@@ -0,0 +1,115 @@
+-module(lobby).
+-export([init/3]).
+%%Version: 1.0
+%%Authors: Sebastian Fransson, Tim Jonasson
+
```
Comment (pontuslaestadius) [inline Server/src/lobby.erl]: A small header explanation of what the module does overall would be great!

File: Server/src/lobby.erl [hunk 33]
```diff
@@ -0,0 +1,115 @@
...
+  Format_result = io_lib:format("~p", [{Diagram_ID, Class_names}]) ++ "~",
+  gen_tcp:send(Socket, Format_result),
+  send_classes(Socket, Diagrams).
+  
+%Finds the correct diagram from the given list
+find_diagram(_, []) -> not_created;
+find_diagram(Diagram_id,[{Diagram_id, _ ,Pid} | _]) -> Pid;
+find_diagram(Diagram_id,[_| Diagrams])  -> find_diagram(Diagram_id, Diagrams).  
+  
+%Finds the given member in the given list of members
```
Comment (pontuslaestadius) [inline Server/src/lobby.erl:109]: The name is correct but I think it's to explicit. Since the functionality is 'find'. _member is not required. More generic = better!

File: Server/src/lobby.erl [hunk 34]
```diff
@@ -0,0 +1,115 @@
...
+	{create_diagram, Creator_Socket, {Did, Class_names, Classes, Messages}} -> 
+	  %Sends the class names and messages to the client
+      %The character ~ is used as the stop character for when the client should stop reading from the tcp connection
+      Diagram_ID = "l" ++ integer_to_list(Lobby_ID) ++ "d" ++ integer_to_list(Did),
+	  Format_result = io_lib:format("~p", [{Diagram_ID, Class_names}]) ++ "~",
+      send_messages(Members, Format_result),
+	  Self = self(),
+	  loop(Creator_Socket, Password, Members, [{Diagram_ID, Class_names, spawn(fun () -> diagramcoordinator:init(Self, Diagram_ID, {Classes, Messages}) end)}| Diagrams], Lobby_ID);
+	
+	%This case happens when the message recieved contains a command for simulation the diagram
```
Comment (pontuslaestadius) [inline Server/src/lobby.erl]: simulating*

File: Server/src/lobby.erl [hunk 35]
```diff
@@ -0,0 +1,115 @@
...
+	{command, Creator_Socket, {Did, Message_request}} -> 
+	  Pid = find_diagram(Did, Diagrams),
+	  %Sends the message to the diagram coordinator.
+	  Pid ! {Message_request, self()},
+	  %Confirmation that the diagram coordinator received the message
+	  receive
+	  	ok -> ok
```
Comment (pontuslaestadius) [inline Server/src/lobby.erl]: Maybe make sure that the diagram coordinator was the process that sent the message? Or is that hard to implement?

File: Server/src/lobby.erl [hunk 36]
```diff
@@ -0,0 +1,115 @@
...
+	  Format_result = io_lib:format("~p", [{Diagram_ID, Class_names}]) ++ "~",
+      send_messages(Members, Format_result),
+	  Self = self(),
+	  loop(Creator_Socket, Password, Members, [{Diagram_ID, Class_names, spawn(fun () -> diagramcoordinator:init(Self, Diagram_ID, {Classes, Messages}) end)}| Diagrams], Lobby_ID);
+	
+	%This case happens when the message recieved contains a command for simulation the diagram
+	{command, Creator_Socket, {Did, Message_request}} -> 
+	  Pid = find_diagram(Did, Diagrams),
+	  %Sends the message to the diagram coordinator.
+	  Pid ! {Message_request, self()},
```
Comment (pontuslaestadius) [inline Server/src/lobby.erl]: Cool!

File: Server/src/lobby.erl [hunk 37]
```diff
@@ -0,0 +1,115 @@
...
+	  send_messages(Members, [Format_result ++ "~"]),
+	  loop(Creator_Socket, Password, Members, Diagrams, Lobby_ID);
+	  
+	%This case happens when a message has been sent in a diagram and the info is supposed to be sent to the client
+	{message_sent, Did, From, To, Message, Message_number} -> 
+	  %Turns the result into binary
+	  %The character ~ is used as the stop character for when the client should stop reading from the tcp connection
+	  Format_result = io_lib:format("~p", [{Did, From, To, Message, Message_number}]) ++ "~",
+      %Sends it to the client
+	  send_messages(Members, [Format_result]),
```
Comment (pontuslaestadius) [inline Server/src/lobby.erl]: This is inconsistant with all the other send_messages calls. They end with ++ "~" within the function call.

File: Server/src/lobby.erl [hunk 38]
```diff
@@ -0,0 +1,115 @@
...
+	  %Turns the result into binary
+	  Format_result = io_lib:format("~p", [{Did, simulation_finished, Message_number}]),
+      %Sends it to the client
+	  send_messages(Members, [Format_result ++ "~"]),
+	  loop(Creator_Socket, Password, Members, Diagrams, Lobby_ID);
+	  
+	%This case happens when a message has been sent in a diagram and the info is supposed to be sent to the client
+	{message_sent, Did, From, To, Message, Message_number} -> 
+	  %Turns the result into binary
+	  %The character ~ is used as the stop character for when the client should stop reading from the tcp connection
```
Comment (pontuslaestadius) [inline Server/src/lobby.erl:82]: This is a weird place to put this  'help' message?

File: Server/src/lobby.erl [hunk 39]
```diff
@@ -0,0 +1,115 @@
...
+find_diagram(Diagram_id,[{Diagram_id, _ ,Pid} | _]) -> Pid;
+find_diagram(Diagram_id,[_| Diagrams])  -> find_diagram(Diagram_id, Diagrams).  
+  
+%Finds the given member in the given list of members
+find_member([], _) -> not_found;
+find_member([Member|_], Member) -> found_member;
+find_member([_|Members], Member) -> find_member(Members, Member).
+
+%Sends a message to all the users in the given list
+send_messages(Members, Message) -> [gen_tcp:send(Socket, Message) || Socket <- Members].
```
Comment (pontuslaestadius) [inline Server/src/lobby.erl:115]: **Since 'all' (or most) calls to this method look like this:**

```
Format_result = io_lib:format("~p", [{Did, simulation_finished, Message_number}]),
send_messages(Members, [Format_result ++ "~"]),
```

**I suggest just revamping the function to include this functionality.**

```
send_messages(Members, [{Did, atom, Number}]) -> 
	send_messages(Members, [io_lib:format("~p", [{Did, atom, Number}] ++ "~"]);
send_messages(Members, Message) -> 
	[gen_tcp:send(Socket, Message) || Socket <- Members]. 
```

*** This has not been tested, and is just a suggestion to simplify many of the cases in the class.**

File: Server/src/lobbycoordinator.erl [hunk 40]
```diff
@@ -0,0 +1,100 @@
...
+  
+%the loop keeps track of the rooms and handles lobby requests.
+loop(Rooms, Lobby_increment) ->
+  receive 
+    %Creates a lobby and saves the monitor reference as well as lobby info to the list of rooms.
+    {Creator_Socket, create_lobby, Password} -> 
+	  % Spawn the process and save the Pid.
+	  Pid = spawn(fun () -> lobby:init(Creator_Socket, Password, Lobby_increment) end),
+	  % Save the reference so that we can use it later for demonitoring.
+      Ref = monitor(process, Pid),
```
Comment (pontuslaestadius) [inline Server/src/lobbycoordinator.erl:18]: indentation is ripped

File: Server/src/lobbycoordinator.erl [hunk 41]
```diff
@@ -0,0 +1,100 @@
...
+	  % Send a confimation to the client saying that the lobby was created.
+	  gen_tcp:send(Creator_Socket, io_lib:format("INFO# Created lobby with id:,  ~p", [Lobby_increment]) ++ "~"),
+	  loop([{Lobby_increment, Password, Pid, Creator_Socket, Ref}|Rooms], Lobby_increment + 1);
+	
+	%Demonitors, kills and removes the host's lobby from the list of rooms.
+	{Creator_Socket, remove_lobby} ->
```
Comment (pontuslaestadius) [inline Server/src/lobbycoordinator.erl:24]: I will just assume that this functionality works.
Because there are 4 lines in a row doing list stuff.

**Review by pontuslaestadius — CHANGES_REQUESTED** [review]
Many of these are merely SUGGESTIONS and not change requests. And I would like a response too. That's why It's a block and not a comment.

File: Application/src/model/Menu.java (hunk 29 above)
Comment (timjon) [reply Application/src/model/Menu.java]: I would say that it fits better here since the coupling would get increased if we move it. 
Its using a lot of things from the model

File: Server/src/lobby.erl (hunk 36 above)
Comment (timjon) [reply Server/src/lobby.erl]: anything that I need to know about?

File: Server/src/lobby.erl (hunk 38 above)
Comment (timjon) [reply Server/src/lobby.erl:82]: pontuslaestadius on 24 Oct 
"Why are you using the tilde symbol? Explain." You are the one that requested that comment

File: Server/src/lobbycoordinator.erl (hunk 40 above)
Comment (timjon) [reply Server/src/lobbycoordinator.erl:18]: That seems to be an issue with github, looks normal in the accual code

File: Application/src/model/Menu.java (hunk 17 above)
Comment (timjon) [reply Application/src/model/Menu.java:280]: Any comment?

File: Server/src/lobby.erl (hunk 38 above)
Comment (pontuslaestadius) [reply Server/src/lobby.erl:82]: I'm refering to the positioning, As it's been used 3 times earlier in the same file without any note. So just put it on the first instance it's used.

File: Application/src/model/Menu.java [hunk 42]
```diff
@@ -28,6 +42,11 @@
     private static Button button_next;
     private static Button button_previous;
     private static Button button_auto;
+    private static Button button_join_lobby;
+    private static Button button_create_lobby;
+    private static Button button_remove_lobby;
+    private static Button button_import_lobby;
```
Comment (kosara-g) [inline Application/src/model/Menu.java:48]: import lobby sounds like you import the lobby itself, it's clearer below, though

File: Server/src/lobby.erl [hunk 43]
```diff
@@ -0,0 +1,116 @@
...
+
+	%This happens when the message received contains a sequence diagram
+	{create_diagram, Creator_Socket, {Did, Class_names, Classes, Messages}} -> 
+	  %Sends the class names and messages to the client
+      %The character ~ is used as the stop character for when the client should stop reading from the tcp connection
+      Diagram_ID = "l" ++ integer_to_list(Lobby_ID) ++ "d" ++ integer_to_list(Did),
+	  Format_result = io_lib:format("~p", [{Diagram_ID, Class_names}]) ++ "~",
+      send_messages(Members, Format_result),
+	  Self = self(),
+	  loop(Creator_Socket, Password, Members, [{Diagram_ID, Class_names, spawn(fun () -> diagramcoordinator:init(Self, Diagram_ID, {Classes, Messages}) end)}| Diagrams], Lobby_ID);
```
Comment (kosara-g) [inline Server/src/lobby.erl:53]: line break much?

File: Server/src/lobbycoordinator.erl (hunk 40 above)
Comment (Sebastian-Fransson) [reply Server/src/lobbycoordinator.erl:18]: ^+1

File: Server/src/lobby.erl (hunk 33 above)
Comment (Sebastian-Fransson) [reply Server/src/lobby.erl:109]: but isn't it better to be more percise in the naming? since the function is finding a member of the lobby.  Just having 'find' could be confusing and present the question "find what?".

File: Server/src/lobby.erl (hunk 38 above)
Comment (timjon) [reply Server/src/lobby.erl:82]: added more of them

File: Server/src/lobby.erl (hunk 39 above)
Comment (timjon) [reply Server/src/lobby.erl:115]: This will be a bit too time consuming to change for now, we can change it if we ever get the time

File: Server/src/lobby.erl (hunk 43 above)
Comment (timjon) [reply Server/src/lobby.erl:53]: Once again github

File: Application/Application.iml (hunk 27 above)
Comment (Sebastian-Fransson) [reply Application/Application.iml:14]: added back the deletetions. Just keep in mind IntelliJ keeps taking them away when doing changes.

**Review by pontuslaestadius — APPROVED** [review]
